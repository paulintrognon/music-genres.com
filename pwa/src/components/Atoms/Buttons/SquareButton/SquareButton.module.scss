.container {
  display: inline-flex;
  background-color: var(--color-tertiary);
  justify-content: center;
  align-items: center;
  box-shadow: 0 9px 10px 0 rgba(0,0,0,.23);
  padding: 1rem 3rem;
  transition: background-color .2s,border .2s,color .2s;
  text-transform: uppercase;
  font-family: Montserrat;
  font-weight: 600;
  font-size: 1.1rem;
  width: auto;

  img {
    margin-right: 0.5rem;
  }
  
  &:hover {
    background-color: var(--color-tertiary-light);
  }
}
