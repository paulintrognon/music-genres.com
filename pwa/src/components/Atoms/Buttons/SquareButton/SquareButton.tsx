import classnames from 'classnames'
import Loader from '../../Loader/Loader'
import classes from './SquareButton.module.scss'

interface Props extends React.ComponentProps<'span'> {
  isLoading?: boolean
}
const SquareButton: React.FC<Props> = ({ isLoading, className, children, ...props }) => {
  return (
    <span className={classnames(classes.container, className)} {...props}>
      {isLoading ? <Loader /> : children}
    </span>
  )
}
export default SquareButton
