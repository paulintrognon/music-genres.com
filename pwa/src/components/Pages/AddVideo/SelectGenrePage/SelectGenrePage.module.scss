.container {
  width: 38rem;
  max-width: 100%;
  margin: auto;
}

.addGenreContainer {
  margin-top: 2rem;
  text-align: center;
}
