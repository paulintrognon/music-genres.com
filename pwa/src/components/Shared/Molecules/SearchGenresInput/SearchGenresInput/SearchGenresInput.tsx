import { ChangeEvent, useState } from 'react'
import useSWR from 'swr'
import { GenreSuggestionsApiResultType } from '../../../../../types/Api/GenreSuggestionsApiResultType'
import MagnifyingGlassIcon from '../../../Atoms/Icons/MagnifyingGlass/MagnifyingGlass'
import InputWithIcon from '../../InputWithIcon/InputWithIcon'
import GenresSuggestions from '../GenresSuggestions/GenresSuggestions'
import classes from './SearchGenresInput.module.scss'

interface Props {
  slugToHref: (slug: string) => string
  onChange?: (entry: string) => void
  onSubmit?: (entry: string) => void
  onDirty?: () => void
  onClean?: () => void
}
const SearchGenresInput: React.FC<Props> = ({
  slugToHref,
  onChange,
  onSubmit,
  onDirty,
  onClean,
}) => {
  /**
   * On string query changes, we update query state
   */
  const [typedText, setTypedText] = useState('')
  const handleOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    setTypedText(event.target.value)
    onChange && onChange(event.target.value)
  }

  /**
   * On query state changes, we fetch up to 5 genres suggestions
   */
  const { data, isValidating } = useSWR<GenreSuggestionsApiResultType>(
    typedText && `/api/music-genres/search?query=${typedText}&limit=5`
  )

  /**
   * On submit, we pass the typed text to onSubmit handler (if exists)
   */
  const handleOnSubmit = (): void => {
    onSubmit && onSubmit(typedText)
  }

  return (
    <div className={classes.container}>
      <InputWithIcon
        className={classes.searchInput}
        placeholder="Start looking for a genre"
        icon={<MagnifyingGlassIcon />}
        onDirty={onDirty}
        onClean={onClean}
        onChange={handleOnChange}
        onSubmit={handleOnSubmit}
      />
      {typedText ? (
        <GenresSuggestions
          isLoading={isValidating}
          genreName={typedText}
          suggestions={data?.result || []}
          slugToHref={slugToHref}
        />
      ) : null}
    </div>
  )
}
export default SearchGenresInput
