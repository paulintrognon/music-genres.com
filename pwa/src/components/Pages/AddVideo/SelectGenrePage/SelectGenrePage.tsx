import Link from 'next/link'
import { useRouter } from 'next/router'
import { useState } from 'react'
import { addVideoLink } from '../../../../services/links/links'
import SquareButton from '../../../Atoms/Buttons/SquareButton/SquareButton'
import SearchGenres from '../../../Organisms/SearchGenres/SearchGenres/SearchGenres'
import StepHeading from '../StepHeading/StepHeading'
import classes from './SelectGenrePage.module.scss'

const SelectGenrePage: React.FC = () => {
  const router = useRouter()

  /**
   * Genre Name searched state
   */
  const [genreName, setGenreName] = useState<string>()
  const handleGenreChange = (genre: string): void => {
    setGenreName(genre)
  }

  /**
   * On submit, if a genreName has been entered, we go to next page
   */
  const handleGenreSubmit = (): void => {
    if (genreName) {
      router.push(addVideoLink(genreName))
    }
  }

  return (
    <div className={classes.container}>
      {/* heading */}
      <StepHeading step={1}>Select a genre related to your video</StepHeading>

      {/* Search Genre input bar */}
      <SearchGenres
        slugToHref={addVideoLink}
        onChange={handleGenreChange}
        onSubmit={handleGenreSubmit}
      />

      {/* Submit button (displayed only if genre has been entered) */}
      {genreName ? (
        <div className={classes.addGenreContainer}>
          <Link href={addVideoLink(genreName)}>
            <a>
              <SquareButton>Create &quot;{genreName}&quot;</SquareButton>
            </a>
          </Link>
        </div>
      ) : null}
    </div>
  )
}

export default SelectGenrePage
