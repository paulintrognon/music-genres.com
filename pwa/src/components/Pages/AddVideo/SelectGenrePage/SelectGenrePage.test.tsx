import { shallow } from 'enzyme'
import SelectGenrePage from './SelectGenrePage'

// Mocking userRouter
import { useRouter } from 'next/router'
import SearchGenresInput from '../../../Shared/Molecules/SearchGenresInput/SearchGenresInput/SearchGenresInput'
import SquareButton from '../../../Shared/Atoms/Buttons/SquareButton/SquareButton'
jest.mock('next/router')
const useRouterMock = useRouter as jest.Mock
const routerPushMock = jest.fn()

beforeEach(() => {
  jest.resetAllMocks()
  useRouterMock.mockReturnValue({ push: routerPushMock })
})

it('should render SelectGenrePage', () => {
  const wrapper = shallow(<SelectGenrePage />)
  expect(wrapper).toMatchSnapshot()
})

it('should display genre link on genre typed', () => {
  const wrapper = shallow(<SelectGenrePage />)
  expect(wrapper.find(SquareButton)).toHaveLength(0)

  // On query "test genre" typed ...
  wrapper.find(SearchGenresInput).simulate('change', 'test genre')

  // We expect the button containing the genre to appear
  expect(wrapper.find(SquareButton)).toHaveLength(1)
  expect(wrapper).toMatchSnapshot()
})

it('should go to next page on submit', () => {
  const wrapper = shallow(<SelectGenrePage />)

  // On query "test genre" typed ...
  wrapper.find(SearchGenresInput).simulate('change', 'test genre')
  wrapper.find(SearchGenresInput).simulate('submit')

  // The router must push to the add video page with the typed genre
  expect(routerPushMock.mock.calls).toEqual([['/add/video/test genre']])
})
