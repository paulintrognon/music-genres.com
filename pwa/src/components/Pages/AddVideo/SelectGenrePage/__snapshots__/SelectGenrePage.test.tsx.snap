// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display genre link on genre typed 1`] = `
<div
  className="container"
>
  <StepHeading
    step={1}
  >
    Select a genre related to your video
  </StepHeading>
  <SearchGenresInput
    onChange={[Function]}
    onSubmit={[Function]}
    slugToHref={[Function]}
  />
  <div
    className="addGenreContainer"
  >
    <Link
      href="/add/video/test genre"
    >
      <a>
        <SquareButton>
          Create "
          test genre
          "
        </SquareButton>
      </a>
    </Link>
  </div>
</div>
`;

exports[`should render SelectGenrePage 1`] = `
<div
  className="container"
>
  <StepHeading
    step={1}
  >
    Select a genre related to your video
  </StepHeading>
  <SearchGenresInput
    onChange={[Function]}
    onSubmit={[Function]}
    slugToHref={[Function]}
  />
</div>
`;
